modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/namehandler/FileNameRegexBasedSPI.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * GeoTools - The Open Source Java GIS Toolkit
+ * http://geotools.org
+ *
+ * (C) 2016, Open Source Geospatial Foundation (OSGeo)
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ */
+package org.geotools.gce.imagemosaic.namehandler;
+
+import java.io.File;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.imageio.ImageReader;
+
+import org.apache.commons.io.FilenameUtils;
+import org.geotools.coverage.grid.io.GridCoverage2DReader;
+import org.geotools.gce.imagemosaic.Utils.SourceHandler;
+
+/**
+ * A {@link CoverageNameHandlerSPI} implementation setting up the coverage name based on a regex
+ * applied to the fileName.
+ *
+ */
+public class FileNameRegexBasedSPI implements CoverageNameHandlerSPI {
+
+    public static String REGEX = "regex";
+
+    public CoverageNameHandler create(Object object, Map<String, String> map) {
+        String regex = null;
+        if (map != null && map.containsKey(REGEX) && ((regex = map.get(REGEX)) != null)) {
+            return new FileNameRegexBasedCoverageNameHandler(regex);
+        }
+        throw new IllegalArgumentException("FileNameRegexBasedSPI should have "
+                + "a defined REGEX property in the map");
+    }
+
+    /**
+     * A {@link CoverageNameHandler} implementation which setup the coverageName
+     * on top of a regex on the filename
+     */
+    static class FileNameRegexBasedCoverageNameHandler implements CoverageNameHandler {
+
+        /** regex to extract the coverage name from the file name */ 
+        private String regex;
+
+        /** the coverageName */ 
+        private String coverageName;
+
+        public FileNameRegexBasedCoverageNameHandler(String regex) {
+            this.regex = regex;
+        }
+
+        @Override
+        public CoverageNameHandler handle(GridCoverage2DReader reader) {
+            Object source = reader.getSource();
+            SourceHandler sourceHandler = new SourceHandler(source);
+            final File file = sourceHandler.getFile();
+            if (file == null) {
+                throw new IllegalArgumentException("Unable to retrieve a valid source file"
+                        + " for the specified reader");
+            }
+            String baseName = FilenameUtils.getBaseName(file.getAbsolutePath());
+            Pattern pattern = Pattern.compile(regex);
```

Comment:
It's it better to compile the pattern in the constructor and hold it as a field?